File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        maxProfits = [0] * 2
        minCosts = [float('inf')] * 2
        for i in range(n):
            for j in range(2):
                minCosts[j] = min(minCosts[j], prices[i] - (maxProfits[j-1] if j else 0))
                maxProfits[j] = max(maxProfits[j], prices[i] - minCosts[j])
        return maxProfits[-1]

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int b0 = INT_MIN; // buy state 0
        int s0 = 0; // sell state 0
        int b1 = INT_MIN; // buy state 1
        int s1 = 0; // sell state 1
        for (int i: prices) {
            b0 = max(b0, -i);
            s0 = max(s0, b0 + i);
            b1 = max(b1, s0 - i);
            s1 = max(s1, b1 + i);
        }
        return s1;
    }
};